lib/api/bucketPutVersioning.js

```diff
@@ -116,10 +118,22 @@ function bucketPutVersioning(authInfo, request, log, callback) {
             return next(null, bucket, versioningConfiguration);
         }),
         (bucket, versioningConfiguration, next) => {
+            // check if replication is enabled if versioning is being suspended
+            const replicationConfig = bucket.getReplicationConfiguration();
+            const replicationEnabled = replicationConfig
+                && replicationConfig.rules
+                && replicationConfig.rules.some(r => r.enabled);
+            if (versioningConfiguration.Status === 'Suspended' &&
+                replicationEnabled) {
+                next(errors.InvalidBucketState
+                    .customizeDescription(invalidBucketStateMessage));
+                return;
+            }
```

Comment:
I would just merge all these checks into one variable or one if condition instead of doing 3 parts in a variable and then another one in the if condition. Considering all of the replication config work is irrelevant if versioning configuration status is not suspended as it would early evaluate in the if block.

```suggestion
            const replicationConfig = bucket.getReplicationConfiguration();
            if (versioningConfiguration.Status === 'Suspended' &&
                replicationConfig &&
                replicationConfig.rules &&
                replicationConfig.rules.some(r => r.enabled)) {
                next(errors.InvalidBucketState
                    .customizeDescription(invalidBucketStateMessage));
                return;
            }
```

Can't wait for optional chaining though...
```javascript
if (versioningConfiguration.Status === 'Suspended' &&
    bucket.getReplicationConfiguration()?.rules?.some(r => r.enabled)) {
```

Reply:
Thanks, I collapsed it into a single conditional. Once we upgrade to Node 14 we will have optional chaining.